### ISSUE TITLE
Algorithm documentation

### ISSUE THREAD

**Issue by MobyGamer** [gligli/soundchunks#1]
Is there documentation on the algorithm somewhere?  I was trying to figure out how the algorithm works by looking at the decompressor, but it wasn't clear (for example, is "chunk" a frame, or subgroupings within a frame?)  Also, lines in the decompressor like: `(b + Low(ShortInt)) * 2047 div High(ShortInt)` are not clear what they do, or why.

**Comment by gligli** [comment]
Ok, I'll try to comment the decoder more and add an explanation of how the different elements are laid out!

In short, the file format is only a bunch of frames, with a header each. Inside are chunks (which are short samples of the same size and have an integer attenuation), and chunk indexes (with some attributes like 'negative' and 'reversed').
The decoder outputs sound by parsing the chunk indexes list, and outputing corresponding attenuated chunks corrected with the attributes.

This line of code converts chunks samples from unsigned 8bits to signed 12bits.

**Comment by MobyGamer** [comment]
Thanks, looking forward to any documentation or explanations.

I understand what the line of code does, but not what low(shortint) and high(shortint) evaluate to.  shortint is a pascal range of -127..128 so maybe those are the low/high values?

OT, but can the format and process be extended to 8-bit PCM?  Currently the algorithm appears competitive with ADPCM in performing a 5:1 reduction of 16-bit PCM data.  Can a similar percentage reduction of 8-bit PCM data be done with similar sonic results?

**Comment by gligli** [comment]
I documented the format in the readme.md!

Yes, + Low(ShortInt) is - 128.

Actually, it could work, regardless of input format by storing chunks in 4 bit PCM but the quality would be nowhere near hifi.
I made the encoder to be able to decode with very low CPU ressources and 8 bit sampling something that is of good enough quality to be like MP3, but without the decoding overhead (eg. on Amiga it would work nicely, with the attenuations being decoded to Paula channels volumes).

**Comment by MobyGamer** [comment]
Documentation looks good, thanks very much.

If there is a way to describe in a few sentences roughly how the algorithm works from the compression side (in the documentation), that would be great.  Based on the format, it appears to be a series of deltas with varying quantization (you use the term "attenuation") applied, but the description of the algorithm mentions machine learning, the encoder has low and high filter thresholds, etc. so there is probably more going on?

Asking because I'm confused about how everything works -- you mention Amiga as a possible output target, however the decoder produces 16-bit output (by "amiga" I'm assuming you intend ganging channels for the pseudo-14-bit Paula playback method).

My area of research focuses around realtime decompression on low-resource platforms (even slower than the 68000) whose output devices are typically capable of 1-bit, 4-bit (log), 6-bit, or 8-bit DAC playback.  I've come up with methods that have acceptable fidelity ranging from 2.6:1 to 6:1 compression of the source 8-bit PCM data, but yours is novel, hence my trying to understand it :-)

**Comment by gligli** [comment]
Cool!

Basically, each frame of audio (a few seconds) is cut into chunks (typically 4 samples each), then reduced to a small number of chunks using K-Means clustering. Then, a K-Nearest neighbors method is used to map those back to the original chunks, giving the chunk indexes. The format is really basically a LUT of chunks with attributes.

On amiga, you would just play the 8bit chunk (given Paula can reverse and flip samples, I'm not sure, else it would mean having 4 versions of the chunk). Attenuations would be converted to the channel volume, so only 2x 8bits channels for near-hifi playback.

Ok cool :) Are you by chance the dude who made the audio and video decoder on the C64 that played boys boys boys by sabrina?

**Comment by MobyGamer** [comment]
Ah, so this is vector quantization of the source material (which is indeed one of the techniques I've explored).  Thanks, that clears things up incredibly (you may want to add that to the documentation/readme).  That does map to 8-bit PCM nicely.  The "fun" part is trying to figure out the best method of VQ (k-means may produce the best local optimums but for audio you need something that addresses global optimums).  I improved my process by adding "simulated annealing" to the VQ process, if you're curious.

If you're producing new codebooks every second or so, that helps quality, so kudos for doing that.

I'm not Algorithm / Onslaught (real name: Naveem) who did the C64 demos you mention, but I've exchanged ideas with him many times.  I'm Trixter / Hornet (one of the team members who produced 8088 MPH and Area 5150).

**Comment by gligli** [comment]
Yeah, I tried many clustering methods already, and at least K-Means is fast and not bad at all. I wonder how you use annealing for clustering...

Yeah, one big codebook isn't very nice when in comes to packing the format.

Ok, my mistake, and Kudos for Area 5150, unbelievable stuff :)
